src/main/java/fr/xephi/authme/api/v3/AuthMePlayerImpl.java

```diff
@@ -0,0 +1,92 @@
+package fr.xephi.authme.api.v3;
+
+import fr.xephi.authme.data.auth.PlayerAuth;
+
+import java.time.Instant;
+import java.util.UUID;
+
+/**
+ * Implementation of {@link AuthMePlayer}. This implementation is not part of the API and
+ * may have breaking changes in subsequent releases.
+ */
+class AuthMePlayerImpl implements AuthMePlayer {
+
+    private String name;
+    private UUID uuid;
+    private String email;
+
+    private Instant registrationDate;
+    private String registrationIpAddress;
+
+    private Instant lastLoginDate;
+    private String lastLoginIpAddress;
+
+    AuthMePlayerImpl() {
+    }
+
+    /**
+     * Maps the given player auth to an AuthMePlayer instance. Returns null if
+     * the player auth is null.
+     *
+     * @param playerAuth the player auth or null
+     * @return the mapped player auth, or null if the argument was null
+     */
+    static AuthMePlayer fromPlayerAuth(PlayerAuth playerAuth) {
+        if (playerAuth == null) {
+            return null;
+        }
+
+        AuthMePlayerImpl authMeUser = new AuthMePlayerImpl();
+        authMeUser.name = playerAuth.getRealName();
+        authMeUser.uuid = playerAuth.getUuid();
+        authMeUser.email = nullIfDefault(playerAuth.getEmail(), PlayerAuth.DB_EMAIL_DEFAULT);
+        Long lastLoginMillis = nullIfDefault(playerAuth.getLastLogin(), PlayerAuth.DB_LAST_LOGIN_DEFAULT);
+        authMeUser.registrationDate = toInstant(playerAuth.getRegistrationDate());
+        authMeUser.registrationIpAddress = playerAuth.getRegistrationIp();
+        authMeUser.lastLoginDate = toInstant(lastLoginMillis);
+        authMeUser.lastLoginIpAddress = nullIfDefault(playerAuth.getLastIp(), PlayerAuth.DB_LAST_IP_DEFAULT);
+        return authMeUser;
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    public UUID getUuid() {
+        return uuid;
+    }
+
+    @Override
+    public String getEmail() {
+        return email;
+    }
+
+    @Override
+    public Instant getRegistrationDate() {
+        return registrationDate;
+    }
+
+    @Override
+    public String getRegistrationIpAddress() {
+        return registrationIpAddress;
+    }
+
+    @Override
+    public Instant getLastLoginDate() {
+        return lastLoginDate;
+    }
+
+    @Override
+    public String getLastLoginIpAddress() {
```

Comment:
Is there a reason we don't use InetAddresses? I know they perform a DNS request, but only on DNS names not IP addresses.

Reply:
Most cases I'm thinking of would compare the IP address with another string IP from elsewhere, or store it somewhere, no? I'm worried it will be a more annoying type than helpful as I'd anticipate a lot of uses having to unwrap it. But if you think otherwise I'm happy to change it.

(Also I'm lazy because constructing InetAddresses might throw UnknownHostException, which should never happen, but we'd still have to handle them properly :laughing:). It's not an excuse to not do it and as mentioned, I'm happy to do it if you really want.

Another alternative is offering both `String` and `InetAddress` but I think we should be careful with our interface(s), i.e. choose an approach and kind of stick with it. Say, if we added more fields to the AuthMePlayer it would quickly become confusing.

Reply:
>Also I'm lazy because constructing InetAddresses might throw UnknownHostException,

Well if it's correct IP address we could ignore it. If not, it doesn't make sense to return this invalid string either.

>Most cases I'm thinking of would compare the IP address with another string IP from elsewhere, or store it somewhere, no? I'm worried it will be a more annoying type than helpful as I'd anticipate a lot of uses having to unwrap it. But if you think otherwise I'm happy to change it.

Ok I agree with you. I think the methods of `InetAddress` are most likely not relevant over the `String` use cases.